torch/distributed/_sharded_tensor/__init__.py

```diff
@@ -48,3 +49,39 @@ def empty(
         pin_memory=pin_memory,
         memory_format=memory_format,
         process_group=process_group)
+
+def state_dict_hook(module, destination, prefix, local_metadata):
```

Comment:
why we need to have a static_dict_hook to update the state_dict? could we directly adding the sharded tensor to `nn.module.state_dict`? I feel like it's weird that user need to register a hook in order to save a sharded tensor, ideally user would not notice it as sharded tensor is just another form of tensor.

Reply:
I couldn't find a better way to add ShardedTensor to the state_dict. ShardedTensor is not a subclass of Tensor so something like `torch.nn.Parameter(my_sharded_tensor)` doesn't work. Also, if you look at `torch.nn.Parameter` it calls `_make_subclass` which detaches the tensor and does a bunch of Tensor specific calls under the hood. Even if we make ShardedTensor to be a subclass of Tensor it is not clear what this Tensor is since ShardedTensor can actually have a list of local shards.

The PyTorch core team is aware of the problem and they are planning to design a brand new API to support this in the future: https://github.com/pytorch/pytorch/issues/62094. But for now, it looks like this hook approach is the best way forward.